packages/telemetry/src/flight-recorder.js

```diff
@@ -206,31 +181,212 @@ export const makeMemoryMappedCircularBuffer = async ({
       );
     }
 
+    writeRecord(record, firstWriteLength, circEnd);
+
+    header.setBigUint64(
+      I_CIRC_END,
+      (circEnd + BigInt(record.byteLength)) % arenaSize,
+    );
+  };
+
+  return { readCircBuf, writeCircBuf };
+}
+
+// TODO use .write() instead of .writeSync() because the writes are already
+// buffered earlier in the slog sender infra, and fsync after every block
+/**
+ * Variant of makeMemoryMappedCircularBuffer that writes to a file instead of using BufferFromFile
+ *
+ * @param {{
+ *   circularBufferSize?: number,
+ *   stateDir?: string,
+ *   circularBufferFilename?: string
+ * }} opts
+ * @returns {Promise<CircularBuffer>}
+ */
+export const makeSimpleCircularBuffer = async ({
+  circularBufferSize = DEFAULT_CBUF_SIZE,
+  stateDir = '/tmp',
+  circularBufferFilename,
+}) => {
+  const filename = circularBufferFilename || `${stateDir}/${DEFAULT_CBUF_FILE}`;
+
+  const newArenaSize = await initializeCircularBuffer(
+    filename,
+    circularBufferSize,
+  );
+
+  const file = await fsp.open(filename, 'r+');
+
+  const headerBuffer = Buffer.alloc(I_ARENA_START);
+
+  await file.read({
+    buffer: headerBuffer,
+    length: I_ARENA_START,
+    position: 0,
+  });
+  const header = new DataView(headerBuffer.buffer);
+
+  // Detect the arena size from the header, if not initialized.
+  const hdrArenaSize = header.getBigUint64(I_ARENA_SIZE);
+  const arenaSize = newArenaSize || hdrArenaSize;
+
+  const hdrMagic = header.getBigUint64(I_MAGIC);
+  SLOG_MAGIC === hdrMagic ||
+    Fail`${filename} is not a slog buffer; wanted magic ${SLOG_MAGIC}, got ${hdrMagic}`;
+  arenaSize === hdrArenaSize ||
+    Fail`${filename} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`;
+
+  /** @type {(outbuf: Uint8Array, readStart: number, firstReadLength: number) => void} */
+  const readRecord = (outbuf, readStart, firstReadLength) => {
+    fs.readSync(file.fd, outbuf, {
+      length: firstReadLength,
+      position: Number(readStart) + I_ARENA_START,
+    });
+    if (firstReadLength < outbuf.byteLength) {
+      fs.readSync(file.fd, outbuf, {
+        offset: firstReadLength,
+        length: outbuf.byteLength - firstReadLength,
+        position: I_ARENA_START,
+      });
+    }
+  };
+
+  /**
+   * Writes to the file, offset by the header size
+   *
+   * @param {Uint8Array} record
+   * @param {number} firstWriteLength
+   * @param {bigint} circEnd
+   */
+  const writeRecord = (record, firstWriteLength, circEnd) => {
+    fs.writeSync(
+      file.fd,
+      record.subarray(0, firstWriteLength),
+      // TS saying options bag not available
+      undefined,
+      undefined,
+      Number(circEnd) + header.byteLength,
+    );
+    if (firstWriteLength < record.byteLength) {
+      // Write to the beginning of the arena.
+      fs.writeSync(
+        file.fd,
+        record.subarray(firstWriteLength, record.byteLength),
+        undefined,
+        undefined,
+        header.byteLength,
+      );
+    }
+  };
```

Comment:
Something like?

```suggestion
     // Write out the updated file header.
     fs.writeSync(
       file.fd,
       headerBuffer,
     );
   };
```

Reply:
ah, thanks. I had something in an earlier iteration and forgot to come back to it. Forthcoming.

I was planning to have a callback for updating the value, but I see now that would write every time through the "overlap > 0n" loop. 

Yours is a good solution imo. I'll add it.